### PR TITLE
Declare cheatcode

### LINKED ISSUE #1454: Adapt `declare` and `deploy_contract` to work with cairo1 felt argument.

### PR DESCRIPTION
**Pull request by cptartur**
Closes #1454

### CODE REVIEW

File: protostar/cheatable_starknet/callable_hint_locals/declare_hint_local.py [hunk 1]
```diff
@@ -1,54 +1,65 @@
...
-        return DeclaredContract(class_hash)
+        return DeclaredContract(class_hash=declared_class.class_hash)
+
+
+def make_contract_class(sierra_contract: str) -> ContractClass:
+    return ContractClass.loads(sierra_contract)
+
+
+def compute_class_hash(casm_contract: str) -> int:
+    return compute_compiled_class_hash(parse_casm(json.loads(casm_contract)))
```
Comment (MaksymilianDemitraszek) [inline protostar/cheatable_starknet/callable_hint_locals/declare_hint_local.py:65]: Those functions seem a bit too anemic, can we avoid it? It is completely understandable what they do from their body. Introducing additional functions brings unnecessary mental overhead when jumping around code

File: protostar/cheatable_starknet/controllers/contracts.py [hunk 2]
```diff
@@ -119,10 +133,48 @@ async def _transform_calldata_to_cairo_data(
                 ) from dt_exc
         return calldata or []
 
+    async def declare_sierra_contract(
```
Comment (MaksymilianDemitraszek) [inline protostar/cheatable_starknet/controllers/contracts.py:136]: Why do we need this one in `PreparedContract` 🤔

File: pyproject.toml [hunk 3]
```diff
@@ -27,7 +27,7 @@ tqdm = "^4.65.0"
 typing-extensions = "^4.3.0"
 tomlkit = "^0.11.6"
 maturin = "^0.14.15"
-web3 = "5.31.4"
+web3 = "6.0.0b4"
```
[the remark this answers is not in the document]
Comment (cptartur) [reply pyproject.toml:30]: No. This was a temporary fix. Should be partially resolved on starknet-0.11 branch now.

File: protostar/cheatable_starknet/controllers/contracts.py (hunk 2 above)
Comment (cptartur) [reply protostar/cheatable_starknet/controllers/contracts.py:136]: It's ContractsController. Diff is just weird.

File: protostar/cheatable_starknet/controllers/contracts.py [hunk 4]
```diff
@@ -103,6 +131,12 @@ async def _transform_calldata_to_cairo_data(
     ) -> CairoData:
         if isinstance(calldata, collections.Mapping):
             contract_class = await self.cheatable_state.get_contract_class(class_hash)
+
+            if isinstance(contract_class, CompiledClass):
+                # TODO add handling
+                raise ValueError("Cair1 contracts do not support data transformation.")
```
Comment (cptartur) [inline protostar/cheatable_starknet/controllers/contracts.py:137]: Do we need better handling here?

File: protostar/cheatable_starknet/controllers/contracts.py [hunk 5]
```diff
@@ -117,10 +151,50 @@ async def _transform_calldata_to_cairo_data(
...
+        )
+
+        with self.cheatable_state.copy_and_apply() as state_copy:
+            await tx.apply_state_updates(
+                state=state_copy, general_config=starknet_config
+            )
+
+        abi = contract_class.abi
+        # TODO figure out how it works
+        # self._add_event_abi_to_state(abi)
```
Comment (cptartur) [inline protostar/cheatable_starknet/controllers/contracts.py:187]: Abis of cairo1 contract's are plain strings which is incompatible with abis we use here. Should the abi related logic be updated (and should it be done in this PR) so it's possible to to `_add_event_abi_to_state`?

File: tests/integration/cairo1_cheatcodes/declare/declare_test.cairo [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+            array_append::<felt>(ref data, x);
+            panic(data)
+        },
+    }
+}
+
+#[test]
+fn test_declare_nonexistent() {
+    match declare('abcdef') {
+        Result::Ok(class_hash) => {
```
Comment (MaksymilianDemitraszek) [inline tests/integration/cairo1_cheatcodes/declare/declare_test.cairo:18]: Can we add tests
- declare cairo0 contract - should fail?
- declare broken cairo1 contract

File: tests/integration/cairo1_cheatcodes/declare/declare_test.cairo [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+            assert(class_hash != 0, 'class_hash != 0');
+        },
+        Result::Err(x) => {
+            let mut data = array_new::<felt>();
```
Comment (MaksymilianDemitraszek) [inline tests/integration/cairo1_cheatcodes/declare/declare_test.cairo:22]: Can we assert error message content here instead?

File: tests/integration/cairo1_cheatcodes/declare/data/cairo_project.toml [hunk 8]
```diff
@@ -0,0 +1,2 @@
+[crate_roots]
```
Comment (MaksymilianDemitraszek) [inline tests/integration/cairo1_cheatcodes/declare/data/cairo_project.toml:1]: I don't think contracts need cairo_project, am I right?

File: protostar/compiler/project_compiler.py [hunk 9]
```diff
@@ -104,6 +117,128 @@ def compile_contract_from_contract_identifier(
             )
         return self.compile_contract_from_contract_name(contract_identifier, config)
 
+    def compile_contract_to_sierra_from_contract_identifier(
```
Comment (MaksymilianDemitraszek) [inline protostar/compiler/project_compiler.py:120]: ProjectCompiler is Cairo0 entity, can we rename it to Cairo0Compiler instead and create separate entity for Cairo1?

Comment (MaksymilianDemitraszek) [reply protostar/compiler/project_compiler.py:120]: `compile_contract_to_sierra_from_contract_identifier` shouldn't be supported for cairo1, we should only support `ContractName`

File: protostar/cheatable_starknet/controllers/contracts.py (hunk 4 above)
Comment (MaksymilianDemitraszek) [reply protostar/cheatable_starknet/controllers/contracts.py:137]: It depends it will be raised as Unexpected exception by protostar
If you want exception to be pretty printed you have to raise CheatcodeException

Comment (MaksymilianDemitraszek) [reply protostar/cheatable_starknet/controllers/contracts.py:137]: But it should be catched and re-raised from HintLocal not from Controller for consistency

Comment (MaksymilianDemitraszek) [reply protostar/cheatable_starknet/controllers/contracts.py:137]: @Arcticae

**Comment by MaksymilianDemitraszek** [comment]
Generally LGTM, good job and congrats of first major PR 🔥 left couple comments

File: protostar/compiler/project_compiler.py [hunk 10]
```diff
@@ -37,6 +45,11 @@ class ProjectCompilerConfig:
     hint_validation_disabled: bool = False
 
 
+class OutputType(Enum):
```
Comment (Arcticae) [inline protostar/compiler/project_compiler.py:48]: I don't understand why is that enum used in the compiler, do we need it outside the `ProjectCompiler` class?

**Review by Arcticae — CHANGES_REQUESTED** [review]

File: protostar/cheatable_starknet/controllers/contracts.py (hunk 4 above)
Comment (Arcticae) [reply protostar/cheatable_starknet/controllers/contracts.py:137]: That's right 👍

Comment (cptartur) [reply protostar/cheatable_starknet/controllers/contracts.py:137]: It's not used in Declare cheatcode, so perhaps this should be addressed when updating relevant cheatcodes (deploy_prepared and prepare I believe)?

File: protostar/cheatable_starknet/controllers/contracts.py (hunk 5 above)
Comment (cptartur) [reply protostar/cheatable_starknet/controllers/contracts.py:187]: Should I leave it as todo? I think it's a okay indicator of this feature missing in case someone touches this code.

Comment (MaksymilianDemitraszek) [reply protostar/cheatable_starknet/controllers/contracts.py:187]: It is ok leaving a TODO but please also create and issue

File: tests/integration/cairo1_cheatcodes/declare/declare_test.cairo (hunk 7 above)
Comment (cptartur) [reply tests/integration/cairo1_cheatcodes/declare/declare_test.cairo:22]: We've decided on raising `CheatcodeException` instead. Invalid contract names/paths are non-recoverable errors in tests.

File: protostar/cheatable_starknet/controllers/contracts.py (hunk 5 above)
Comment (cptartur) [reply protostar/cheatable_starknet/controllers/contracts.py:187]: #1663

File: protostar/cheatable_starknet/controllers/contracts.py [hunk 11]
```diff
@@ -138,6 +212,8 @@ async def declare_cairo0_contract(
             )
 
         abi = contract_class.abi
+        assert abi is not None
```
[the remark this answers is not in the document]
Comment (cptartur) [reply protostar/cheatable_starknet/controllers/contracts.py:215]: ABI is now (or was always?) typed as Optional in `DeprecatedCompiledClass`.

File: protostar/compiler/project_compiler.py (hunk 10 above)
Comment (cptartur) [reply protostar/compiler/project_compiler.py:48]: It's not needed outside, I just think using enum here provides stronger typing than relying on strings or other solution. Prefixed it with underscore so it's obvious it's internal.

File: protostar/compiler/project_compiler.py (hunk 9 above)
Comment (Arcticae) [reply protostar/compiler/project_compiler.py:48]: let's remove dead code please

File: protostar/cheatable_starknet/callable_hint_locals/declare_hint_local.py [hunk 12]
```diff
@@ -0,0 +1,71 @@
...
+    ):
+        self._contracts_controller = contracts_controller
+        self._project_compiler = project_compiler
+
+    @property
+    def name(self) -> str:
+        return "declare"
+
+    def _build(self) -> Callable:
+        return self.declare
```
Comment (Arcticae) [inline protostar/cheatable_starknet/callable_hint_locals/declare_hint_local.py:39]: the new style of defining `hint_local`s that was agreed upon is returning a function defined inside of this `_build`

File: protostar/cheatable_starknet/callable_hint_locals/declare_hint_local.py [hunk 13]
```diff
@@ -0,0 +1,71 @@
...
+
+    def declare(self, contract: int) -> DeclaredContract:
```
Comment (Arcticae) [inline protostar/cheatable_starknet/callable_hint_locals/declare_hint_local.py:41]: There shouldn't be this much code in `HintLocal`, can it be moved to the controller? (refer to `DeployHintLocal` as per how much code should be in the hintlocal itself)

File: protostar/cheatable_starknet/controllers/contracts.py [hunk 14]
```diff
@@ -77,6 +85,26 @@ class DeployedContract:
...
+        )
+
+    def run_validate_entrypoint(
+        self,
+        remaining_gas: int,
+        state: SyncState,
+        resources_manager: ExecutionResourcesManager,
+        general_config: StarknetGeneralConfig,
+    ) -> Tuple[Optional[CallInfo], int]:
+        # This hack might interfere with other things
```
Comment (Arcticae) [inline protostar/cheatable_starknet/controllers/contracts.py:104]: If we're doing a hack, can we explain what it hacks? It might help with debugging some day

**Review by Arcticae — CHANGES_REQUESTED** [review]

File: protostar/cheatable_starknet/callable_hint_locals/declare_hint_local.py
[the remark this answers is not in the document]
Comment (MaksymilianDemitraszek) [reply protostar/cheatable_starknet/callable_hint_locals/declare_hint_local.py:1]: @cptartur similarly to old cheatcodes documentation https://docs.swmansion.com/protostar/docs/tutorials/testing/cheatcodes

We can do it in a separate PR to not block anybody

File: protostar/compiler/project_compiler.py
Comment (cptartur) [inline protostar/compiler/project_compiler.py:1]: This works to the extend that is tested by declare hint local tests, but that's all.

Ideally this module should have it own tests (perhaps we can just copy and adapt Cairo0ProjectCompiler tests?). 
Should this be done in this or separate PR (and if in separate, should we prioritize testing this further for in the current milestone?)

@Arcticae @MaksymilianDemitraszek

**Review by Arcticae — APPROVED** [review]

File: protostar/compiler/project_compiler.py
Comment (MaksymilianDemitraszek) [reply protostar/compiler/project_compiler.py:1]: I wouldn't do it at all, I would spend the time instead to remove this abstraction, it should be faster than writing tests

File: protostar/cheatable_starknet/controllers/contracts.py [hunk 15]
```diff
@@ -73,6 +97,31 @@ class DeployedContract:
...
+        #
+        # By default InternalDeclare has a quite complicated run_validate_entrypoint method
+        # that among other things verifies if the sender_address of the Declare transaction is of a deployed contract.
+        # There's no account in protostar internals so this validation will always fail. This hack completely removes
```
Comment (MaksymilianDemitraszek) [inline protostar/cheatable_starknet/controllers/contracts.py:120]: Please use docstring instead

File: protostar/cheatable_starknet/controllers/contracts.py [hunk 16]
```diff
@@ -55,11 +68,22 @@ class ConstructorInvocationException(ContractsCheaterException):
     pass
 
 
+class OperationNotSupportedForCairo1Exception(ProtostarException):
+    def __init__(self, message: str = "Operation not supported for Cairo1 contracts."):
+        super().__init__(message=message)
```
Comment (MaksymilianDemitraszek) [inline protostar/cheatable_starknet/controllers/contracts.py:73]: I would raise CheatcodeException there instead, is there any reason why we need to create a new error type here?

File: protostar/cheatable_starknet/controllers/contracts.py (hunk 15 above)
Comment (MaksymilianDemitraszek) [reply protostar/cheatable_starknet/controllers/contracts.py:120]: but it is a nice use of comments, very informative ❤️

**Review by MaksymilianDemitraszek — APPROVED** [review]

File: protostar/cheatable_starknet/controllers/contracts.py [hunk 17]
```diff
@@ -113,10 +170,51 @@ async def _transform_calldata_to_cairo_data(
...
+        self,
+        contract_class: ContractClass,
+        compiled_class: CompiledClass,
+    ) -> DeclaredSierraClass:
+        """
+        Declare a sierra contract.
+
+        @param contract_class: sierra compiled contract to be declared
+        @param compiled_class_hash: compiled class hash of the casm compiled contract.
+        @param compiled_class: casm compiled contract to be declared
```
Comment (MaksymilianDemitraszek) [inline protostar/cheatable_starknet/controllers/contracts.py:183]: Why we have the same `@param` repeated twice, it is some python docsting convention for alternative?

File: protostar/cheatable_starknet/controllers/contracts.py (hunk 16 above)
Comment (cptartur) [reply protostar/cheatable_starknet/controllers/contracts.py:73]: The previous suggestion would be to raise a specific error here so it can be handled by the cheatcodes. But I guess it doesn't make sense since it cannot be reasonably handled.

File: protostar/cheatable_starknet/controllers/contracts.py (hunk 17 above)
Comment (cptartur) [reply protostar/cheatable_starknet/controllers/contracts.py:183]: Just my error

File: protostar/cheatable_starknet/callable_hint_locals/declare_hint_local.py [hunk 18]
```diff
@@ -0,0 +1,77 @@
...
+        def _compile_contract(
+            contract_identifier: ContractIdentifier,
+        ) -> Tuple[CompiledClass, ContractClass]:
+            contract_class = self._project_compiler.compile_contract_to_sierra_from_contract_identifier(
+                contract_identifier
+            )
+            compiled_class = self._project_compiler.compile_contract_to_casm_from_contract_identifier(
+                contract_identifier
+            )
+            return compiled_class, contract_class
```
Comment (karol-bisztyga) [inline protostar/cheatable_starknet/callable_hint_locals/declare_hint_local.py:75]: is there a good reason to put these two functions inside of the `_build` function? I think they should be just members of the class, they don't capture anything (ok, `declare` captures `_compile_contract` but it could call it with `self.`) and I think nesting like this is not the best practice.

File: protostar/configuration_file/configuration_file.py [hunk 19]
```diff
@@ -106,6 +106,7 @@ def __init__(self, contract_name: str, expected_declaration_location: str):
         super().__init__(
             f"Couldn't find `{contract_name}` in `{expected_declaration_location}`"
         )
+        self.contract_name = contract_name
```
Comment (karol-bisztyga) [inline protostar/configuration_file/configuration_file.py:109]: what is this for?

File: protostar/cheatable_starknet/callable_hint_locals/declare_hint_local.py (hunk 18 above)
Comment (cptartur) [reply protostar/cheatable_starknet/callable_hint_locals/declare_hint_local.py:75]: I think they can be made separate members.

**Comment by MaksymilianDemitraszek** [comment]
@cptartur Can we prioritize merging this one? It is a heavy blocker 🥺

File: protostar/configuration_file/configuration_file.py (hunk 19 above)
Comment (cptartur) [reply protostar/configuration_file/configuration_file.py:109]: I'm re-raising a `CheatcodeException` from `ContractNameNotFoundException` in declare hint local. I need a contract name for the exception message.

File: protostar/cheatable_starknet/callable_hint_locals/declare_hint_local.py (hunk 18 above)
Comment (cptartur) [reply protostar/cheatable_starknet/callable_hint_locals/declare_hint_local.py:75]: We can't make this class methods due to #1693 and `_compile_contract` needs to capture self to use `contracts_controller`. So I don't think these can be made just normal functions.

Comment (karol-bisztyga) [reply protostar/cheatable_starknet/callable_hint_locals/declare_hint_local.py:75]: Discussed offline, we can follow up in the issue #1693

File: protostar/configuration_file/configuration_file.py (hunk 19 above)
Comment (karol-bisztyga) [reply protostar/configuration_file/configuration_file.py:109]: right, thanks

**Review by karol-bisztyga — APPROVED** [review]
please fix the CI (related #1694)
